### PR TITLE
CODENVY-601: Do not run new backup/restore process when another one is still running

### PR DESCRIPTION
**Pull request by mmorhun**
This PR prevents simultaneously run of several backup/restore processes for a workspace

### CODE REVIEW

**Comment by mmorhun** [comment]
@garagatyi @evoevodin @skabashnyuk @sleshchenko review please

File: plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java [hunk 1]
```diff
@@ -89,7 +120,20 @@ public void backupWorkspace(final String workspaceId,
     public void backupWorkspaceAndCleanup(final String workspaceId,
                                           final String srcPath,
                                           final String srcAddress) throws ServerException {
-        backupWorkspace(workspaceId, srcPath, srcAddress, true);
+        ReentrantLock lock = workspacesBackupLocks.get(workspaceId);
+        if (lock != null) {
+            lock.lock();
+            if (workspacesBackupLocks.get(workspaceId) == null) {
+                // should never happen
```
Comment (garagatyi) [inline plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java]: Please describe in comment possible situation when it is possible as in https://github.com/codenvy/codenvy/pull/425/files#diff-6fa2b3319d5dc836fd809bbba1ed5605R95

File: plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java [hunk 2]
```diff
@@ -89,7 +120,20 @@ public void backupWorkspace(final String workspaceId,
...
+                LOG.error("Backup with cleanup of the workspace {} was invoked several times simultaneously", workspaceId);
+            }
+            try {
```
Comment (garagatyi) [inline plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java]: I think It is convenient to put try block right after locking. Please do that to not confuse possible code readers.

File: plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java [hunk 3]
```diff
@@ -89,7 +120,20 @@ public void backupWorkspace(final String workspaceId,
     public void backupWorkspaceAndCleanup(final String workspaceId,
                                           final String srcPath,
                                           final String srcAddress) throws ServerException {
-        backupWorkspace(workspaceId, srcPath, srcAddress, true);
+        ReentrantLock lock = workspacesBackupLocks.get(workspaceId);
+        if (lock != null) {
+            lock.lock();
+            if (workspacesBackupLocks.get(workspaceId) == null) {
+                // should never happen
+                LOG.error("Backup with cleanup of the workspace {} was invoked several times simultaneously", workspaceId);
```
Comment (garagatyi) [inline plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java]: Should we skip backup and cleanup in that case?

File: plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java [hunk 4]
```diff
@@ -131,6 +177,13 @@ public void restoreWorkspaceBackup(final String workspaceId,
                                        final String userId,
                                        final String groupId,
                                        final String destAddress) throws ServerException {
+        ReentrantLock lock = new ReentrantLock();
+        lock.lock();
+        if (workspacesBackupLocks.putIfAbsent(workspaceId, lock) != null) {
+            LOG.error("Attempt to start backup/restore process on {} workspace while it under restore.", workspaceId);
+            return; // we cannot restore workspace while it under backup/restore process
```
Comment (garagatyi) [inline plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java]: Comment is confusing a bit. Please rephrase it to explain that this code identify multiple restores of one workspace which is illegal.

File: plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java [hunk 5]
```diff
@@ -131,6 +177,13 @@ public void restoreWorkspaceBackup(final String workspaceId,
...
+        }
+
         try {
```
Comment (garagatyi) [inline plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java]: Please put try block right after locking to not confuse readers of code.

File: plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java (hunk 4 above)
Comment (sleshchenko) [inline plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java]: mb it would be better to unlock lock in this case

Comment (garagatyi) [inline plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java]: Sure, but if try block was first code after locking this problem would not appear.

File: plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/WorkspaceFsBackupScheduler.java [hunk 6]
```diff
@@ -76,12 +80,13 @@ public void scheduleBackup() {
                     isTimeToBackup(machineId)) {
 
                     executor.execute(() -> {
-                        try {
-                            backupWorkspaceInMachine(machine);
-
-                            lastMachineSynchronizationTime.put(machine.getId(), System.currentTimeMillis());
-                        } catch (ServerException | NotFoundException e) {
-                            LOG.error(e.getLocalizedMessage(), e);
```
Comment (garagatyi) [inline plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/WorkspaceFsBackupScheduler.java]: Is this logging dropped on purpose?

File: plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/WorkspaceFsBackupScheduler.java [hunk 7]
```diff
@@ -76,12 +80,13 @@ public void scheduleBackup() {
                     isTimeToBackup(machineId)) {
 
                     executor.execute(() -> {
-                        try {
-                            backupWorkspaceInMachine(machine);
```
Comment (garagatyi) [inline plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/WorkspaceFsBackupScheduler.java]: Looks like actual backup execution is missing

File: plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java (hunk 3 above)
Comment (mmorhun) [inline plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java]: yes, of course

File: plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java [hunk 8]
```diff
@@ -89,7 +120,22 @@ public void backupWorkspace(final String workspaceId,
...
+                    // should never happen
+                    LOG.error("Backup with cleanup of the workspace {} was invoked several times simultaneously", workspaceId);
+                    return;
+                }
+                backupWorkspace(workspaceId, srcPath, srcAddress, true);
+            } finally {
+                workspacesBackupLocks.remove(workspaceId);
+                lock.unlock();
+            }
+        }
```
Comment (garagatyi) [inline plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java]: Please add the same as in https://github.com/codenvy/codenvy/pull/425/files#diff-6fa2b3319d5dc836fd809bbba1ed5605R106

**Comment by sleshchenko** [comment]
LGTM

**Comment by garagatyi** [comment]
LGTM

**Comment by skabashnyuk** [comment]
ok

**Comment by garagatyi** [comment]
@riuvshin Can we merge that PR now?

**Comment by riuvshin** [comment]
@garagatyi  if it is ready lets merge it
